ml/Database.cc

```diff
@@ -38,6 +38,78 @@ const char *ml::Database::SQL_SELECT_ANOMALY_EVENTS =
     "   after >= ?4 AND"
     "   before <= ?5;";
 
+const char *ml::Database::SQL_CREATE_ANOMALY_RATE_INFO_TABLE =
+    "CREATE TABLE IF NOT EXISTS anomaly_rate_info( "
+    "     host_id text NOT NULL, "
+    "     after int NOT NULL, "
+    "     before int NOT NULL, "
+    "     anomaly_rates text "
+    ");";
+
+const char *ml::Database::SQL_INSERT_BULK_ANOMALY_RATE_INFO =
+    "INSERT INTO anomaly_rate_info( "
+    "     host_id, after, before, anomaly_rates) "
+    " VALUES (?1, ?2, ?3, ?4);";
+   
+/*const char *ml::Database::SQL_SELECT_ANOMALY_RATE_INFO =
+    "SELECT dimension_id, AVG(anomaly_percentage) "
+    "  FROM (SELECT json_extract(j.value, '$[1]') AS dimension_id, "
+    "  json_extract(j.value, '$[0]') AS anomaly_percentage "
+    "  FROM anomaly_rate_info AS ari, json_each(ari.anomaly_rates) AS j " 
+    "  WHERE ari.host_id == ?1 AND ari.after >= ?2 AND ari.before <= ?3 "
+    "  AND json_valid(ari.anomaly_rates)) "
+    "  GROUP BY dimension_id;";*/
+
+ const char *ml::Database::SQL_SELECT_ANOMALY_RATE_INFO =
```

Comment:
Considering the complexity of the query we need to add unit tests that verify its correctness. You can check Manolis' approach in https://github.com/netdata/netdata/pull/11794/files to get an idea how this would work. In our case, I think it would be better to have a separate, standalone `.sql` file that will populate a table with some data and run a couple queries that will verify the correctness of this statement.

Another issue we need to address is the query's performance. It's very difficult to predict how well it will behave in a real setting. I think we need to:

1. Install somewhere the agent with ML enabled,
2. Let it run until DB engine starts rotating data, and
3. Manually run various queries for large time ranges.

By doing the 3rd step we can see how the system behaves and how fast we can get back the data. However, we need to discuss this in the sync meeting in order to see if the rest of the team agrees with this approach.

Reply:
OK, as you know, we have not yet taken profiling steps, as if it starts serving at staging level, we could do so. I agree with the Unit Test, my manual tests were approving to fair extent, but will do the unit test. in terms of complexity, I would not expect exhaustive situations, but we will see :)